### PR TITLE
Add fcntl subcommand F_GETOWN/F_SETOWN

### PR DESCRIPTION
**Pull request by cqs21**
see issue: #960

### CODE REVIEW

**Comment by tatetian** [comment]
@liqinggd Could you review this PR?

File: kernel/aster-nix/src/fs/file_table.rs [hunk 1]
```diff
@@ -274,6 +330,7 @@ impl Clone for FileTableEntry {
             file: self.file.clone(),
             flags: AtomicU8::new(self.flags.load(Ordering::Relaxed)),
             subject: Subject::new(),
+            owner: RwLock::new(None),
```
Comment (liqinggd) [inline kernel/aster-nix/src/fs/file_table.rs:333]: I have a question, when cloning fd, will the owner be duplicated?

Comment (cqs21) [reply kernel/aster-nix/src/fs/file_table.rs:333]: Yes, you're right. I'll fix it.

**Comment by liqinggd** [comment]
LGTM

File: kernel/aster-nix/src/fs/file_table.rs [hunk 2]
```diff
@@ -232,6 +238,39 @@ pub struct FileTableEntry {
...
     flags: AtomicU8,
     subject: Subject<FdEvents>,
+    owner: RwLock<Option<Owner>>,
+}
+
+type Owner = (Pid, Arc<dyn Observer<IoEvents>>);
+
+struct OwnerObserver {
+    file: Arc<dyn FileLike>,
+    owner: Pid,
```
Comment (tatetian) [inline kernel/aster-nix/src/fs/file_table.rs:248]: I think it is more reliable to keep a weak reference to the owner process. The owner process may exit and its pid will be re-assigned to other processes. In this case, the signal may be sent to the wrong process.

```rust
impl Observer<IoEvents> for OwnerObserver {
    fn on_events(&self, events: &IoEvents) {
        if self.file.status_flags().contains(StatusFlags::O_ASYNC)
            && let Some(process) = process_table::get_process(self.owner)
```

Furthermore, I/O events may occur quite frequently. So the `on_events` method should be implemented as efficient as possible. `process_table::get_process` will have to lock the entire process table. So it is relatively costly. Instead, if the `OwnerObserver` keeps a weak reference to the owner process, then the `on_events` method only needs to do `Weak::upgrade`, which is much cheaper than `on_events`.

File: kernel/aster-nix/src/fs/file_table.rs [hunk 3]
```diff
@@ -240,13 +279,30 @@ impl FileTableEntry {
...
+
+    pub fn set_owner(&self, owner: Pid) -> Result<()> {
+        if let Some((_, observer)) = self.owner.read().as_ref() {
+            let _ = self.file.unregister_observer(&Arc::downgrade(observer));
+        }
+
+        let observer = OwnerObserver::new(self.file.clone(), owner);
+        self.file
+            .register_observer(observer.weak_self(), IoEvents::empty())?;
+        *self.owner.write() = Some((owner, observer));
```
Comment (tatetian) [inline kernel/aster-nix/src/fs/file_table.rs:302]: This method is not implemented in a thread-safe way as the lock-protected scope is too small. Think about what would happen if `set_owner` is called concurrently.

File: kernel/aster-nix/src/fs/file_table.rs [hunk 4]
```diff
@@ -240,13 +279,30 @@ impl FileTableEntry {
...
 
     pub fn file(&self) -> &Arc<dyn FileLike> {
         &self.file
     }
 
+    pub fn owner(&self) -> Option<Pid> {
+        self.owner.read().as_ref().map(|(pid, _)| *pid)
+    }
+
+    pub fn set_owner(&self, owner: Pid) -> Result<()> {
```
Comment (tatetian) [inline kernel/aster-nix/src/fs/file_table.rs:294]: The meaning of this so-called owner desires documentation. You don't even know the meaning before you write this PR, right? Other people are no smarter than you. So please write down what you learn as comments for future readers, including a six-month-later you.

**Comment by tatetian** [comment]
The code looks good. But it is not shipped with any test case. Are there such test cases in the gvisor test suite?

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**